src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/UserAgentHelper.java

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+package com.microsoft.signalr;
+
+public class UserAgentHelper {
+
+    private final static String USER_AGENT = "User-Agent";
+
+    public static String getUserAgentName() {
+        return USER_AGENT;
+    }
+
+    public static String createUserAgentString() {
+        StringBuilder agentBuilder = new StringBuilder("Microsoft SignalR/");
+
+        // Parsing version numbers
+        String detailedVersion = Version.getDetailedVersion();
+        agentBuilder.append(getVersion(detailedVersion));
+        agentBuilder.append("; ");
+        agentBuilder.append(detailedVersion);
+        agentBuilder.append("; ");
+
+        // Getting the OS name
+        agentBuilder.append(findOSName(System.getProperty("os.name")));
+        agentBuilder.append("; Java; ");
+
+        // Vendor and Version
+        agentBuilder.append(System.getProperty("java.vendor"));
+        agentBuilder.append("; ");
+        agentBuilder.append(System.getProperty("java.version"));
+
+        return agentBuilder.toString();
+    }
+
+    static String getVersion(String detailedVersion) {
+        String version;
+        if (detailedVersion.contains("-")) {
+            version = detailedVersion.substring(0, detailedVersion.indexOf('-'));
+        } else {
+            version = detailedVersion;
+        }
+        return version;
+    }
+
+    static String findOSName(String operatingSystem) {
+        operatingSystem = operatingSystem.toLowerCase();
```

Comment:
We should see what this does on Android. If possible, it would be awesome to put in `Android` as the OS name and the actual Android version for the OS version.

Reply:
Just did some searching and it seems like there are some other ways (java.vendor for example) to check if you're running on Android but I'm curious to see what all the different values are. I'll spin up Android studio and check it out.

Reply:
How ugly would it be if we just included the os.name property directly in the user agent?

Reply:
![AndroidDebugValues](https://user-images.githubusercontent.com/4809660/63298559-9143ef00-c288-11e9-8ec9-bec32c0cc751.png)

Reply:
The vendor seems like the best way to determine if we're running on Android. FWIW Stackoverflow agrees.

Reply:
I agree that we should include both os and vendor info in the UA. The question I have is why do we need the findOSName at all instead of appending os.name directly?

In the Android case, it seems that os.name is already "Linux" so findOSName does nothing making this a moot point. But what if someone really is running on AIX or something weird, but it just shows "Linux" in the user agent? I'd rather not lose information unless os.name is sometimes extremely ugly and/or verbose.

Reply:
I don't really feel strongly either way. I understand the desire to standardize in case it's verbose like you mentioned but we could also just use the value directly. I'm fine with either, just don't want to get hung up on this.

Reply:
An argument for standardizing could be it makes it easier to test

Reply:
My vote is for including the unaltered os.name. The tests can reference os.name and just make sure it matches and is in the right spot. I agree it's not a big deal though.